spec/API_specification/linear_algebra_functions.md

```diff
@@ -11,8 +11,74 @@ A conforming implementation of the array API standard must provide and support t
+
+5.  **Orthogonality**: an interface should have clearly defined and delineated functionality which, ideally, has no overlap with the functionality of other interfaces in the specification. Providing multiple interfaces which can all perform the same operation creates unnecessary confusion regarding interface applicability (i.e., which interface is best at which time) and decreases readability of both library and user code. Where overlap is possible, the specification must be parsimonious in the number of interfaces, ensuring that each interface provides a unique and compelling abstraction. Examples of related interfaces which provide distinct levels of abstraction (and generality) include:
+
+    -   `vecdot`: computing the dot product of two vectors.
+    -   `matmul`: performing matrix multiplication (including between two vectors and thus the dot product).
+    -   `tensordot`: computing tensor contractions (generalized sum-products).
+    -   `einsum`: expressing operations in terms of Einstein summation convention, including dot products and tensor contractions.
+
+    The above can be contrasted with, e.g., NumPy, which provides the following interfaces for computing the dot product or related operations:
+
+    -   `dot`: dot product, matrix multiplication, and tensor contraction.
+    -   `inner`: dot product.
+    -   `vdot`: dot product with flattening and complex conjugation.
+    -   `multi_dot`: chained dot product.
+    -   `tensordot`: tensor contraction.
+    -   `matmul`: matrix multiplication (dot product for two vectors).
+    -   `einsum`: Einstein summation convention.
+
+    where `dot` is overloaded based on input array dimensionality and `vdot` and `inner` exhibit a high degree of overlap with other interfaces. By consolidating interfaces and more clearly delineating behavior, this specification aims to ensure that each interface has a unique purpose and defined use case.
+
 ## Objects in API
 
+```{note}
+
+Historically, linear algebra functionality has been split across interfaces belonging to different namespaces. For example, NumPy exposes linear algebra interfaces in both its top-level namespace and its `linalg` sub-namespace. The determination of which linear algebra interface belongs to which namespace is often unintuitive and a common source of confusion.
```

Comment:
What exists in `numpy.linalg` can be understood as based on upon what requires a matrix factorization and what doesn't, i.e., LAPACK vs BLAS.

This is arguably a pretty important distinction for both users and array libraries as well. The matrix factorization stuff is much harder to implement and has more niche use-cases for scientific computing. In contrast, the stuff that doesn't (even including matmul) is easier to implement and useful even for more generic data analysis use-cases, e.g., computing a weighted average. Libraries like xarray and dask would use/implement only the non-factorization stuff.

`matmul` is a bit of a gray area (a naive loop particularly inefficient, compared to an optimized implementations), but I think there's a strong case for including it in the main namespace for a few reasons:
1. It's a really fundamental building block. Every array library implements it and you need it both for almost all scientific computing use-cases and also deep learning. Even data analysis libraries like xarray use it (though perhaps they don't really _need_ it).
2. The algorithms for efficient matmul (at least in practice) tend to be relatively simple in contrast to matrix factorization -- the main question is how to implement the method efficiently, not what method can be used. The efficiency trade-off is more similar to the use of SIMD instructions for vectorized arithmetic than a matrix factorization.

Reply:
> This is arguably a pretty important distinction for both users and array libraries as well.

I would say for libraries it is, for users it is not. Users don't care about how easy something is to implement. And saying, e.g., that `tensordot` or `einsum` is more important or fundamental than `svd` or `qr` is (imho) wrong. Only physicists/mathematicians for example tend to know Einstein notation - and they'd certainly also know and use `svd` and `qr`.

I think that's the main issue I have with the current split in NumPy. API design driven from implementation details up tends to be suboptimal.

Reply:
One idea could be to have the `linalg` namespace contain both the factorization set and the linalg functions in the main namespace. That'd be much clearer to users. For all regular array libraries that anyway ship with BLAS/LAPACK: only use the `linalg` namespace, everything is there.